packages/react-devtools-scheduling-profiler/src/hooks.js

```diff
@@ -0,0 +1,199 @@
+  useEffect(() => {
+    if (modalRef.current === null) {
+      return () => {};
+    }
+
+    const handleDocumentKeyDown = ({key}: any) => {
+      if (key === 'Escape') {
+        dismissCallback();
+      }
+    };
+
+    const handleDocumentClick = (event: any) => {
+      // $FlowFixMe
+      if (
+        modalRef.current !== null &&
+        !modalRef.current.contains(event.target)
+      ) {
+        event.stopPropagation();
+        event.preventDefault();
+
+        dismissCallback();
+      }
+    };
+
+    // It's important to listen to the ownerDocument to support the browser extension.
+    // Here we use portals to render individual tabs (e.g. Profiler),
+    // and the root document might belong to a different window.
+    const ownerDocument = modalRef.current.ownerDocument;
+    ownerDocument.addEventListener('keydown', handleDocumentKeyDown);
+    if (dismissOnClickOutside) {
+      ownerDocument.addEventListener('click', handleDocumentClick);
+    }
+
+    return () => {
+      ownerDocument.removeEventListener('keydown', handleDocumentKeyDown);
+      ownerDocument.removeEventListener('click', handleDocumentClick);
+    };
+  }, [modalRef, dismissCallback, dismissOnClickOutside]);
+}
+
+// Copied from https://github.com/facebook/react/pull/15022
+export function useSubscription<Value>({
+  getCurrentValue,
+  subscribe,
+}: {|
+  getCurrentValue: () => Value,
+  subscribe: (callback: Function) => () => void,
+|}): Value {
+  const [state, setState] = useState({
+    getCurrentValue,
+    subscribe,
+    value: getCurrentValue(),
+  });
+
+  if (
+    state.getCurrentValue !== getCurrentValue ||
+    state.subscribe !== subscribe
+  ) {
+    setState({
+      getCurrentValue,
+      subscribe,
+      value: getCurrentValue(),
+    });
+  }
+
+  useEffect(() => {
+    let didUnsubscribe = false;
+
+    const checkForUpdates = () => {
+      if (didUnsubscribe) {
+        return;
+      }
+
+      setState(prevState => {
+        if (
+          prevState.getCurrentValue !== getCurrentValue ||
+          prevState.subscribe !== subscribe
+        ) {
+          return prevState;
+        }
+
+        const value = getCurrentValue();
+        if (prevState.value === value) {
+          return prevState;
+        }
+
+        return {...prevState, value};
+      });
+    };
+    const unsubscribe = subscribe(checkForUpdates);
+
+    checkForUpdates();
+
+    return () => {
+      didUnsubscribe = true;
+      unsubscribe();
+    };
+  }, [getCurrentValue, subscribe]);
+
+  return state.value;
+}
+
+export function useBrowserTheme(): BrowserTheme {
+  const [theme, setTheme] = useState<BrowserTheme>('light');
+
+  useEffect(() => {
+    const handlePreferredColorSchemeChange = (event: MediaQueryListEvent) => {
+      setTheme(event.matches ? 'dark' : 'light');
+    };
+
+    const mediaQueryList = window.matchMedia('(prefers-color-scheme: dark)');
+    mediaQueryList.addEventListener('change', handlePreferredColorSchemeChange);
```

Comment:
I don't think this picks up on dark mode if it's already set (only if the user changes their system theme *after* the profiler has been opened). At least that seems to be the behavior on OSX (the "change" event won't get fired).

We could *also* check `window.matchMedia(DARK_MODE_QUERY).matches` in the effect and update, but then we might have to render twice (once for the default "light" and then again for "dark").

I think this might be a nice opportunity to use the `useMutableSource` hook! :smile: It will handle all of this for us. Here's how you'd use it:

```js
import {
  unstable_createMutableSource as createMutableSource,
  unstable_useMutableSource as useMutableSource,
  useEffect,
  useState,
} from 'react';

export type BrowserTheme = 'dark' | 'light';

const DARK_MODE_QUERY = '(prefers-color-scheme: dark)';

const getSnapshot = window =>
  window.matchMedia(DARK_MODE_QUERY).matches ? 'dark' : 'light';

const darkModeMutableSource = createMutableSource(
  window,
  () => window.matchMedia(DARK_MODE_QUERY).matches,
);

const subscribe = (window, callback) => {
  const mediaQueryList = window.matchMedia(DARK_MODE_QUERY);
  mediaQueryList.addEventListener('change', callback);
  return () => {
    mediaQueryList.removeEventListener('change', callback);
  };
};

export function useBrowserTheme(): BrowserTheme {
  return useMutableSource(darkModeMutableSource, getSnapshot, subscribe);
}
```

Reply:
Wow, cool! I didn't know this existed 😆

Reply:
It's not released yet in stable, but DevTools uses prerelease APIs :smile:

https://github.com/reactjs/rfcs/blob/master/text/0147-use-mutable-source.md